src/treelearner/feature_histogram.hpp

```diff
@@ -670,35 +771,42 @@ class FeatureHistogram {
     return Common::Sign(s) * reg_s;
   }
 
-  template <bool USE_L1, bool USE_MAX_OUTPUT>
+  template <bool USE_L1, bool USE_MAX_OUTPUT, bool USE_SMOOTHING>
   static double CalculateSplittedLeafOutput(double sum_gradients,
                                             double sum_hessians, double l1,
-                                            double l2, double max_delta_step) {
+                                            double l2, double max_delta_step,
+                                            double smoothing, data_size_t num_data,
+                                            double parent_output) {
+    double ret;
     if (USE_L1) {
-      double ret = -ThresholdL1(sum_gradients, l1) / (sum_hessians + l2);
+      ret = -ThresholdL1(sum_gradients, l1) / (sum_hessians + l2);
       if (USE_MAX_OUTPUT) {
         if (max_delta_step > 0 && std::fabs(ret) > max_delta_step) {
-          return Common::Sign(ret) * max_delta_step;
+          ret = Common::Sign(ret) * max_delta_step;
         }
       }
-      return ret;
     } else {
-      double ret = -sum_gradients / (sum_hessians + l2);
+      ret = -sum_gradients / (sum_hessians + l2);
       if (USE_MAX_OUTPUT) {
         if (max_delta_step > 0 && std::fabs(ret) > max_delta_step) {
-          return Common::Sign(ret) * max_delta_step;
+          ret = Common::Sign(ret) * max_delta_step;
         }
       }
-      return ret;
     }
+    if (USE_SMOOTHING) {
+      ret = ret * (num_data / smoothing) / (num_data / smoothing + 1) \
```

Comment:
maybe only apply this when `smoothing > 0` ?

Reply:
I have done the check for `smoothing > kEpsilon` outside this function (e.g. in `FuncForNumricalL2` and `FuncForCategoricalL1`) , and I only pass the template parameter `USE_SMOOTHING=true` if `smoothing > kEpsilon`. Alternatively we could move the check inside `CalculateSplittedLeafOutput`, and then we wouldn't need the template parameter `USE_SMOOTHING`. This would be simpler but we would lose the speedup that the template provides.

Reply:
`USE_SMOOTHING` may be useless for this function. the soothing output can be simplified as below。
weight for current output is `w = n/(n+s)` and `ret = w * ret + (1-w) * parent_out`. `n` must be greater zero. `s=0` means no path smoothing.